Visualizers properly represent complex numbers

sumitpuri opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #3640 by adding logic for complex imaginary numbers.


### Details and comments

I have added a fix which includes:
(1) Test whether the input is a complex imaginary number. 
(2) If it is, the numerical part of the imaginary number is extracted and made to run the same logic as for a floating number. 
(3) In the end the string is returned by appending 'j' to the numerical part achieved after computation.

claassistantio commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=3666) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=3666) before we can accept your contribution.<br/><hr/>**sumit.puri@us.ibm.com** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=3666) it.</sub>

1ucian0 commented on qiskit/visualization/tools/pi_check.py:
```diff
@@ -44,13 +44,20 @@ def pi_check(inpt, eps=1e-6, output='text', ndigits=5):
         return str(inpt)
     elif isinstance(inpt, str):
         return inpt
+    elif isinstance(inpt, complex) and inpt.real == 0:
```
why `inpt.real == 0`?

sumitpuri replied:
I think I had an incorrect logic in mind that real numbers were being handled by the regular float. This code worked only in scenarios where either of real and imaginary is non-zero. I have now modified the code to take into account both real and imaginary as non-zero values.

1ucian0 commented on qiskit/visualization/tools/pi_check.py:
```diff
@@ -124,5 +131,9 @@ def pi_check(inpt, eps=1e-6, output='text', ndigits=5):
         return str_out
     # nothing found
-    str_out = '%.{}g'.format(ndigits) % inpt
+    if complex_check == 'True':
```
The `isinstance(inpt, complex)` from line 53 could be moved here and that would avoid this `complex_check` flag, right?

sumitpuri replied:
This alone would have not worked because the code "inpt = float(inpt)" appears way above on the top and any complex number as inpt gets errored out at this stage.

sumitpuri commented:
> I think the workflow is a bit too complicated... Maybe the all the numbers after line 53 should be handle as tuples (real, img)?

I  will agree with this. I have modified the code to process the tuples of real and imaginary numbers.

sumitpuri commented:
I did perform make style, make lint, and make test 
Not sure what is causing these checks to fail above.

sumitpuri commented:
@1ucian0 Could you please suggest the next action item on how to resolve these checks that are failing. Thanks.

1ucian0 commented:
I think i fixed the lint issues in https://github.com/Qiskit/qiskit-terra/pull/3666/commits/fbcf3123c1aae55ca1cdd2e51b96a29d79627dff

Now.. some test need readjusting?

1ucian0 commented:
Also, it would be great if you can add like `test_text_pifrac` (in `test/python/visualization/test_circuit_text_drawer.py`) to check the case from #3640 and other that you can come up with.

sumitpuri commented:
> I think i fixed the lint issues in [fbcf312](https://github.com/Qiskit/qiskit-terra/commit/fbcf3123c1aae55ca1cdd2e51b96a29d79627dff)
> 
> In [68b8c3b](https://github.com/Qiskit/qiskit-terra/commit/68b8c3bd9df1663dcfac8eb4bcbf825cfcbf98eb) I did some refactor, based on your code. What do you think?

Refactoring is too good.. condensing the for loop in one map function!! Thank you!

sumitpuri commented:
> Also, it would be great if you can add like `test_text_pifrac` (in `test/python/visualization/test_circuit_text_drawer.py`) to check the case from #3640 and other that you can come up with.

I have added the `test_text_pifrac` function. How can I test this code which I have added in test_circuit_text_drawer.py?

sumitpuri commented:
Seems like I just pushed request to merge 2,209 commits. I was following instructions (https://help.github.com/en/github/using-git/changing-author-info) to clear out my ibm email from the failing CLA check.

sumitpuri commented:
> Dont forget to add some more testing!

Done, added a test where "pi" gets displayed in the text.

1ucian0 commented on releasenotes/notes/fix3640-70e912031815d09a.yaml:
```diff
@@ -0,0 +1,6 @@
+---
+fixes:
+  - |
+    Fixes the circuit to handle complex numbers. Refer to
```
```suggestion
    Fixes the circuit visualization to handle complex numbers.
```
